### DISCUSSION TITLE
Configuration for RAM and CPU

### DISCUSSION THREAD

**Discussion by gcelano** [korpling/ANNIS#867] [General]
My Annis instance (4.12.2) seems to get clogged when I open more windows (sometimes, not even the initial Annis logo appears), as if there are not enough resources available on the server (even if it has 256 CPUs and 200GB for RAM). I have imposed no RAM limit on the service (`annis.service`), but I set up Nice=7 for it (ApacheHTTPServer has `MemoryLimit=104857600000`). No matter how many windows I open, the RAM memory consumption does not increase (while all CPUs are used), and therefore I am not sure whether there is some limit in place within Annis and/or ApacheHTTPServer preventing more/faster access to Annis.  I changed Apache `mpm_event.conf`, but it does not seem to affect anything:

```
StartServers            4
MinSpareThreads         25
MaxSpareThreads         75
ThreadLimit             128
ThreadsPerChild         50
MaxRequestWorkers       300
MaxConnectionsPerChild  0
```

my `service.toml` is
  
```                                                                     
[auth]
anonymous_access_all_corpora = true

[database]
graphannis = "/var/local/annis/.annis/v4"
sqlite = "/var/local/annis/.annis/v4/service_data.sqlite3"
disk_based= false
cache = {PercentOfFreeMemory = 50.0}
```
Do you have any suggestion? Should I change `service.toml` in some respect?

**Comment by thomaskrause** [comment]
Hi, 200GB of RAM seems plenty and even half of it would be used for loading the graph in memory, it should still be enough for the other services.

There can be several bottlenecks, and you already checked Apache and the graphANNIS service. You can check that the graphANNIS service configuration is actually used by the service by looking at the output/log and searching for the line `Starting the bundled graphANNIS service with configuration file /tmp/...`. This shows a path to the actual file used, which might be a modified one derived from the one you configure. You can open it and compare it with the configuration you intended and check that the relevant parts are the same (there might be some differences for the auth configuration).

Another bottleneck can be the interaction between the java-based server that services the web application (which is based on Jetty) and the Apache2 server which acts as a proxy. The Apache2 needs to proxy Websocket requests, which sometimes get interrupted and this looks like the website is not responding anymore, even if the initial frontend page was loaded. You can check with the Firefox or Chrome developer tools if the network connections are not failing. If there is a publicly reachable instance of your server, I can also check if I see any errors from the browser when trying to access it.

P.S. I converted the issue to a discussion, because it is not clear yet where a possible bug in ANNIS might be. We should try to narrow the issue down and then create issues for the specific found bug(s) in ANNIS

**Comment by chiarcos** [comment]
Thanks for this discussion. Is there an info page about the general hardware requirements for ANNIS4 (relative to, say, number of tokens, or (better) markables/nodes in the corpus)? We haven't faced an issue yet, but also worked with smaller corpora, only.

**Reply by thomaskrause** [reply]
Its difficult to estimate the needed main memory space, because it also depends on the structure of the edge component how much memory it will take. E.g. the ordering of the tokens will be stored in a graph storage optimized for such linear graphs (fan-out at most 1) while syntactic trees might be stored in a pre/post-order based graph storage and other structures might fall back to an adjecency list. Also, for main memory based graph storages, we also optimize the data types that are used internally, e.g. only using a 16 bit integer for the pre/post order instead of 32 bit one. If we would just give example corpus data estimations, this could deceive if the scaling is not as expected.

You could use the [`info`](https://korpling.github.io/graphANNIS/docs/v3/cli.html#info) command of the graphANNIS command line interface to check the used corpus resources. In order to create such an information page one might need to configure graphANNIS to a "worst-case" scenario (by using the most pessimistic graph storage memory wise) and then measure the used main memory for a set of known example corpora, where people might have an idea how large  and how complicated the annotation layers are.
